alns/stopping_criteria/StoppingCriterion.py

```diff
@@ -0,0 +1,29 @@
+from abc import ABC, abstractmethod
+
+from numpy.random import RandomState
+
+from alns.State import State
+
+
+class StoppingCriterion(ABC):
+    """
+    Base class from which to implement a stopping criterion.
+    """
+
+    @abstractmethod
+    def __call__(self, best: State, current: State) -> bool:
```

Comment:
Question: does this conform to the signature of `AcceptanceCriterion`?

Reply:
It is slightly different than `AcceptanceCriterion.__call__`:
```python
    @abstractmethod
    def __call__(self,
                 rnd: RandomState,
                 best: State,
                 current: State,
                 candidate: State) -> bool:
```
- I didn't include `rnd:  RandomState` because I couldn't think of a use case where the stopping criterion would involve randomness. Maybe simheuristics?
- I didn't include `candidate: State` because 1) the `candidate` variable is undefined before iteration and 2) candidate solutions do not seem to be used in the evaluation of stopping criteria as listed above.

It is easy to modify the `StoppingCriterion` signature though so that it conforms to the signature of `AcceptanceCriterion`. I leave the decision to you.

Reply:
>I didn't include rnd: RandomState because I couldn't think of a use case where the stopping criterion would involve randomness. Maybe simheuristics?

To be fair: me neither. But the main issues with _not_ having it is that then no stopping criterion can ever depend on the random number stream directly passed to ALNS. That seems restrictive to me, and somewhat defeats the purpose of ALNS accepting a random state in the first place. Even if it never ends up being used, I like the idea that later stopping criteria have the option to use randomness in making their decisions.

I would prefer the signature to take `rnd` as a first argument, like the acceptance criteria.

>I didn't include candidate: State because 1) the candidate variable is undefined before iteration and 2) candidate solutions do not seem to be used in the evaluation of stopping criteria as listed above.

I had not realised it did not yet exist when we call `stop` :-). I agree with leaving the candidate solution out, and stick to best and current.